ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala

```diff
@@ -0,0 +1,1191 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.backend
+
+import java.time.{Duration, Instant}
+
+import com.daml.daml_lf_dev.DamlLf
+import com.daml.ledger.api.domain
+import com.daml.ledger.api.v1.event.{CreatedEvent, ExercisedEvent}
+import com.daml.ledger.participant.state.v1.{
+  ApplicationId,
+  CommandId,
+  Configuration,
+  DivulgedContract,
+  Offset,
+  ParticipantId,
+  Party,
+  RejectionReasonV0,
+  SubmissionId,
+  SubmitterInfo,
+  TimeModel,
+  TransactionId,
+  TransactionMeta,
+  Update,
+  WorkflowId,
+}
+import com.daml.lf.crypto
+import com.daml.lf.data.{Ref, Time}
+import com.daml.lf.ledger.EventId
+import com.daml.lf.transaction.BlindingInfo
+import com.daml.lf.transaction.test.TransactionBuilder
+import com.daml.lf.value.Value
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.appendonlydao.JdbcLedgerDao
+import com.daml.platform.store.appendonlydao.events.Raw.TreeEvent
+import com.daml.platform.store.appendonlydao.events.{
+  CompressionStrategy,
+  ContractId,
+  Create,
+  Exercise,
+  FieldCompressionStrategy,
+  LfValueSerialization,
+  Raw,
+}
+import com.daml.platform.store.dao.DeduplicationKeyMaker
+import com.google.protobuf.ByteString
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class UpdateToDbDtoSpec extends AnyWordSpec with Matchers {
+  // DbDto case classes are not comparable, because the contains Arrays.
+  // The DbDto trait mixes in SomeArrayEquals, but this does not help as some DTOs contain Arrays within Options.
+  // These tests work nevertheless because all array values use the same empty array instance.
```

Comment:
Definitely outside the scope of this PR, but: general advice to use `immutable.ArraySeq` if you need to put arrays in `case class`es, using unsafe array wrapping methods the runtime cost is negligible at most and you don't mess with equality, which is one of the defining characteristics of `case class`es.

Reply:
Thats a nice idea!
...if you want to have your case class comparable that regard, which we don't, seems like referential comparision works all nicely for tests, and no comparision is needed for prod

Reply:
I would rather use `com.daml.lf.data.Bytes`.
If necessary I am fine to move it outside `com.daml.lf.data` and put it in `scala-utils`.

Reply:
> general advice to use `immutable.ArraySeq` if you need to put arrays in `case class`es

In our codebase, use `ImmArray` or `ImmArraySeq` instead (depending on whether you want stdlib compatibility).